File: 239. Sliding Window Maximum.py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        if not nums:
            return []
        x = 1
        q = [float('inf')]
        ret = []
        for i in nums[:k]:
            while len(q) > x and i > q[-1]:
                q.pop()
            q.append(i)
        ret.append(q[x])
        for idx, i in enumerate(nums[k:], k):
            if q[x] == nums[idx-k]:
                x += 1
            while len(q) > x and i > q[-1]:
                q.pop()
            q.append(i)
            ret.append(q[x])
        return ret

File: 239. Sliding Window Maximum.go
func maxSlidingWindow(nums []int, k int) []int {
	n := len(nums)
	res := make([]int, 0, n)
	slid := make([]int, n)
	l, r := 0, 0

	for i := 0; i < n; i++ {
		if i >= k {
			if l < r && nums[i-k] == slid[l] {
				l++
			}
		}
		for l < r && slid[r-1] < nums[i] {
			r--
		}
		slid[r] = nums[i]
		r++
		if i >= k-1 {
			res = append(res, slid[l])
		}
	}
	return res
}
